app/rdf/query-hierarchies.ts

```diff
@@ -26,66 +24,80 @@ const queryDimensionValues = async (dimension: string) => {
   return rows.map((r) => r.value.value);
 };
 
+const getName = (pointer: AnyPointer, language: string) => {
+  let name = pointer.out(ns.schema.name, { language })?.value;
+  if (name) {
+    return name;
+  }
+  return pointer.out(ns.schema.name)?.value;
+};
+
+const toTree = (
+  results: HierarchyNode[],
+  dimensionIri: string,
+  locale: string
+): HierarchyValue[] => {
+  const serializeNode = (
+    node: HierarchyNode,
+    depth: number
+  ): HierarchyValue => {
+    const res: HierarchyValue = {
+      label: getName(node.resource, locale) || "-",
+      value: node.resource.value,
+      children: node.nextInHierarchy.map((childNode) =>
+        serializeNode(childNode, depth + 1)
+      ),
+      depth,
+      dimensionIri: dimensionIri,
+    };
+    return res;
+  };
+  return results.map((r) => serializeNode(r, 0));
+};
+
 export const queryHierarchy = async (
-  dimension: string
+  dimensionIri: string,
+  locale: string
 ): Promise<HierarchyValue[]> => {
-  const query = DESCRIBE`?hierarchy ?level1 ?level2`.WHERE`    
-      SELECT DISTINCT ?hierarchy ?level1 ?level2
-  
-      WHERE {
-    
-        ?cube ${ns.cube.observationConstraint} ?shape.
-        ?shape ?prop ?blankNode.
-        ?blankNode ${ns.sh.path} <${dimension}>.
-        ?blankNode ${ns.cubeMeta.hasHierarchy} ?hierarchy.
-        ?hierarchy ${ns.cubeMeta.hierarchyRoot} ?level1.
-        ?level1 ${ns.schema.hasPart} ?level2 .
-    
-        filter(isiri(?level2))
-    
-        OPTIONAL { ?this ${ns.schema.name} ?order0 } 
-      }
-    `;
+  const source = createSource();
 
-  const dimensionValues = new Set(await queryDimensionValues(dimension));
-  const stream = await query.execute(sparqlClientStream.query);
-  const dataset = await fromStream(rdf.dataset(), stream);
-  const cf = clownface({ dataset });
+  const cubeQuery = SELECT`?cube`.WHERE`
+  ?cube ${ns.cube.observationConstraint} ?shape.
+  ?shape ?prop ?dimension.
+  ?dimension ${ns.sh.path} <${dimensionIri}>.
+  
+  FILTER NOT EXISTS { ?cube ${ns.schema.expires} ?any . }
+  `;
+  const cubeResults = await cubeQuery.execute(sparqlClient.query);
+  if (cubeResults.length === 0) {
+    throw new Error("Could not find cube");
+  }
+  const cubeIri = cubeResults[0].cube.value;
+  const cube = await source.cube(cubeIri);
+  if (!cube) {
+    throw new Error("Could not find cube");
+  }
+  const hierarchy = cube.ptr
+    .any()
```

Comment:
What does this "any" do?

Reply:
Good question, I had not checked the code. I saw this as a way to reset the current pointer to be the whole graph.

The official documentation explains better

> Removes current pointers from the context and return an "any pointer". The returned object can be used to find any nodes in the dataset

https://zazuko.github.io/clownface/#/api?id=clownface%20any

Reply:
Thanks 🙇